Sources/TSCUtility/Archiver.swift

```diff
@@ -0,0 +1,157 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import TSCBasic
+import Dispatch
+
+/// The `Archiver` protocol abstracts away the different operations surrounding archives.
+public protocol Archiver {
+
+    /// A set of extensions the current archiver supports.
+    var supportedExtensions: Set<String> { get }
+
+    /// Asynchronously extracts the contents of an archive to a destination folder.
+    ///
+    /// - Parameters:
+    ///   - archivePath: The `AbsolutePath` to the archive to extract.
+    ///   - destinationPath: The `AbsolutePath` to the directory to extract to.
+    ///   - completion: The completion handler that will be called when the operation finishes to notify of its success.
+    func extract(
+        from archivePath: AbsolutePath,
+        to destinationPath: AbsolutePath,
+        completion: @escaping (Result<Void, Error>) -> Void
+    )
+}
+
+/// An `Archiver` conforming type that stores multiple archivers and selects the one most fitted for each operation.
+public struct ArchiverStore: Archiver {
```

Comment:
This seems like an unnecessary abstraction to me. We can add it later if needed but I'd suggest just adding a simple `ZipArchiver` protocol and then a `CLIZipTool` implementation.

Reply:
Really? I would have thought that would be the very first abstraction we would need. @neonichu, what do you think of this?

Reply:
Hmm? I thought we're only planning on supporting zip archives right now.

Reply:
Oh yes, definitely! I just had the feeling that people would quickly want it to support other archive types, `gzip` for instance. Btw, what is the abstraction that you don't think is necessary? `ArchiverStore` itself or `Archiver`?

Reply:
Hmm, I doubt we'd quickly want to support other archives because that would mean incompatibility with older tools. If we think those are important to support, maybe we should try to propose the popular extensions from get go? I am just talking about getting rid of the `ArchiverStore`.

Reply:
Ok, I trust your judgment. I'll remove `ArchiverStore`.